crates/bevy_utils/src/lib.rs

```diff
@@ -212,6 +212,11 @@ impl Hasher for PassHasher {
         panic!("can only hash u64 using PassHasher");
     }
 
+    #[inline]
+    fn write_u32(&mut self, i: u32) {
+        self.hash = i as u64;
+    }
+
```

Comment:
This is probably a performance footgun. `hashbrown` uses the top 7 bits of the hash to do SIMD probing inside a group, and this will set all of them to 0, essentially nullifying that optimization.

Also note that `Entity` has two `u32` fields, so it will call `write_u32` twice and the second one will overwrite the result of the first one. Luckily the `index` is the second field of `Entity`, but we're not documenting this anywhere and it would be easy to break (especially considering the ongoing work to refactor identifiers in general).

Reply:
Ah, I didn't know it worked like that. What do you suggest?

Reply:
> Ah, I didn't know it worked like that. What do you suggest?

You could try with a very simple hasher like `rustc_hash`. If you want to manually implement it for the `u32` case, it's just `(i as u64).wrapping_mul(0x517cc1b727220a95)`. The only concern I have with this is that it kinda randomizes the hash, so it pessimizes the case where entities are sorted.

An alternative for better memory locality might be keeping the entity index for the lower 32 bits while using the hash for the upper 32 bits, which should allow some useful top 7 bits. This could be implemented with `((i as u64).wrapping_mul(0x517cc1b727220a95) << 32) | (i as u64)`.
For 32 bit targets (e.g. wasm) this is a bit more complicated because `hashbrown` only considers the bottom 32 bits. Unfortunately the top 7 bits of the entity index are all zeros most of the time (you need 2^25 ~ 33M entities for them to start being non-zero). My guess is that you could get away with the previous method but done on a 16-16 bits split, e.g. `(i.wrapping_mul(0x9e3779b9) << 16) | (i & 0xffff)`

> Just noting that the index is the bit I'm interested in using as a key for cases where the map is cleared each frame, in case that makes things simpler to improve.

Then using only `index` should be fine for your uses, but how do you ensure that whoever changes `Entity` in the future makes sure that the `index` field will be hashed last?

Reply:
Yeah I don't like how implicit this is. PassHasher exists to pass a single full u64 hash. Giving it support for a u32 and then using it on _each_ of the fields of Entity is a confusing misuse of that design imo.

Some ideas:
1. A custom Entity Hash impl that combines the index and generation into a single u64
    * Probably very bad for some safety reason (in addition to cleanliness / maintainability), but could we transmute `&self` (`&Entity`) to `&64`?
2. A custom EntityMap type that does the work for users. (ex: if we really want a entity.index->X map, we could just wrap an Entity api over `PassHashMap<u64, X, PassHash>` that casts entity.index to u64.

Reply:
I tried a suggestion from @nicopap about doing a wrapping_mul by u32::MAX / golden ratio. I was thinking that as the hash is a u64, I should instead use u64::MAX / phi. But this made performance really really bad. Not sure what I did wrong. And ahash and others are not as bad as that was, which seems odd. My current guess without having had time to dig deeper is that wrapping_mul on a u64 uses u128 maths that just has really bad performance.

@cart options 1 and 2 don’t address skifire’s comment about the upper 7 bits being used for some simd optimisation. Option 2 would always have the upper 32 bits as all zeros, and option 1 would be mostly zeros unless people do a lot of respawning of entities.

Reply:
Yeah I don't have answers for that :smile: I see my solutions as improvements to the current impl (which also has the zeros problem).

Reply:
I’m happy to take one of your options. There’s also Entity.to_bits().

Reply:
You could copy the 7 bottom bits to the top with a bitshift operation. Regarding the implicitness, I suggested an alternative that is probably equal in performance but much more explicit:

```rust
struct CustomPassHash<T: GetHashValue>(T);

trait GetHashValue { get_hash_value(&self) -> u64 }
impl<T: GetHashValue> Hash for ValueHash<T> {
  fn hash(&self, state: &mut H) {
    state.write_u64(self.0.get_hash_value());
  }
}
```

Reply:
I removed `write_u32`, and implemented `Hash` for `Entity` manually by using `.to_bits()` which shifts the generation into the upper `32 bits`. This seems to work fine, and I don't see why it would be a problem anywhere. Shout if you think of something.

I also tried `FxHasher` (from the `rustc-hash` crate) and `SeaHasher` as they seemed to be claimed to be the fastest hashers I could find, and neither are close to being the speed of past hash. As we are concerned about performance, it seems that the hashing cost of `FxHasher` or similar dominates performance versus the SIMD optimisation in hashbrown's hash map lookup. Maybe we can find something better in the future, but this seems to be fastest now.

Reply:
> You could copy the 7 bottom bits to the top with a bitshift operation.

The problem with this approach is that the 7 top bits are used to differentiate values that end up in the same bucket (or other near buckets), but if they end up in the same bucket their very likely have the same bottom bits, and hence if you copy the 7 bottom bits to the 7 top bits they will be useless to distinguish different values.

Ideally you would copy the bottom 7 bits that aren't included in the bucket mask, but this depends on the size of the hashmap which is not a constant.

> I also tried `FxHasher` (from the `rustc-hash` crate) and `SeaHasher` as they seemed to be claimed to be the fastest hashers I could find, and neither are close to being the speed of past hash. As we are concerned about performance, it seems that the hashing cost of `FxHasher` or similar dominates performance versus the SIMD optimisation in hashbrown's hash map lookup.

Did you try using just those hasher or my version which stores the hash in the top 32 bits only (the one with `((i as u64).wrapping_mul(0x517cc1b727220a95) << 32) | (i as u64)`)?

If you're using the hashers from those crates you may be observing the issue with memory locality I was mentioning previously. And if this is the issue I wonder how often it comes up in practice: this just be a consequence of spawning lot of entities with the same component at the same moment. Maybe the entity distribution is much more random if lot of other systems are running, spawning/despawning entities and changing their archetypes.

Did you hash just the `index` or the generation too? Hashing the generation too takes more time (for example `rustc_hash` is just a multiplication to hash a single `u32`/`u64`, but to hash a second `u32` it starts requiring a rotation, a xor and another multiplication)

> And I did also try .rotate_right(7) to shift the bottom 7 bits up to being the top 7 bits. That made performance seriously tank. Not sure why.

If you `.rotate_right(7)` then you "lose" the bottom 7 bits, which makes every chunk of 128 entities map to the same bucket, which is horrible for performance. I think @nicopap meant doing something like `((i as u64) << 57) (i as u64)` which makes the 7 top bits equal to the 7 bottom bits without losing them, but as I mentioned above this will likely be useless for the SIMD optimization.

Reply:
I tested this PR as-is with Entity.to_bits() and PassHash, versus Skifire's suggestion of using only the index, not the generation, and calculating the hash as `(i as u64).wrapping_mul(0x517cc1b727220a95) << 32) | (i as u64)`, and a suggestion from nicopap to try `i | (i << (64 - 7))` as a cheap way of getting the low 7 bits into the high 7 bits.

## Ordered z spawning

With entities spawned with ordered z (which means after sorting, the batching and main pass 2d are doing a linear scan in terms of entity spawn order):

PassHash (yellow) vs Skifire's (red) frame times:
![Screenshot 2023-09-26 at 15 39 19](https://github.com/bevyengine/bevy/assets/302146/07aef59f-d231-4955-86c6-e7355320f815)

PassHash (yellow) vs nicopap's (red) frame times:
![Screenshot 2023-09-26 at 15 42 50](https://github.com/bevyengine/bevy/assets/302146/d047c1de-49c2-4b2e-b120-7385223ea605)

PassHash mean time per call:
![Screenshot 2023-09-26 at 15 47 38](https://github.com/bevyengine/bevy/assets/302146/f6f8fe0b-3d24-404f-a3de-36828b959056)

Skifire's mean time per call:
![Screenshot 2023-09-26 at 15 48 55](https://github.com/bevyengine/bevy/assets/302146/59b9fab5-bdfc-4ec7-a412-a13ca9649184)

nicopap's mean time per call:
![Screenshot 2023-09-26 at 15 49 13](https://github.com/bevyengine/bevy/assets/302146/db7c33b0-25fe-4577-9bd0-fec456ad9ccb)

nicopap's median frame time is 2.7% less than the PR, and Skifire's is 1.2%.
PR: queuing: 7.43ms, batching: 7ms, main_pass_2d: 0.288ms
nicopap's: queuing: 7.22ms, batching: 6.92ms, main_pass_2d: 0.293ms
Skifire's: queuing: 7.46ms, batching: 6.94ms, main_pass_2d: 0.276ms

So, for ordered, it looks like nicopap's suggestion is marginally better.

## Random z spawning
This mode spawns the entities with random z, this causes their order after sorting to be random, so the lookups in batching and main pass 2d are random.

PassHash (yellow) vs Skifire's (red) frame times:
![Screenshot 2023-09-26 at 15 58 16](https://github.com/bevyengine/bevy/assets/302146/0130e6b8-4e90-45ed-b2ba-fbcfcf4a6901)

PassHash (yellow) vs nicopap's (red) frame times:
![Screenshot 2023-09-26 at 15 58 35](https://github.com/bevyengine/bevy/assets/302146/9f8622a6-05da-4af8-9eff-d659ec7c68da)

PassHash mean time per call:
![Screenshot 2023-09-26 at 15 59 34](https://github.com/bevyengine/bevy/assets/302146/55ab7c65-69a4-471d-bf43-c4b391746f07)

Skifire's mean time per call:
![Screenshot 2023-09-26 at 15 59 50](https://github.com/bevyengine/bevy/assets/302146/0b7dd0b4-5730-4815-9d27-b34d7fe49219)

nicopap's mean time per call:
![Screenshot 2023-09-26 at 16 00 06](https://github.com/bevyengine/bevy/assets/302146/a8ab82cc-bf2d-4986-839a-ba36e36e43b4)

nicopap's median frame time is 2.7% less than the PR, and Skifire's is 3.8%.
PR: queuing: 7.46ms, batching: 25.85ms, main_pass_2d: 88.28ms
nicopap's: queuing: 7.36ms, batching: 24.39ms, main_pass_2d: 86.57ms
Skifire's: queuing: 7.44ms, batching: 23.88ms, main_pass_2d: 85.53ms

So for random z and random lookup orders, Skifire's suggestion is marginally faster.

I suspect Skifire's suggestion is going to be more robust generally, and as it has the best worst case (random access) performance, and is close for the best case (ordered access) performance, I'm inclined to choose it.

Reply:
@SkiFire13 how exactly does one calculate / arrive at `0x517cc1b727220a95`? As I had seen @nicopap's suggestion of using something golden ratio-based, I have seen doing things like u64::MAX / phi. But your constant seems to be u64::MAX / π. However, if I try to calculate it in rust code using `(u64::MAX as f64 * std::f64::consts::FRAC_1_PI) as u64` or with various rounding, I can't get the same number. Is it that such a calculation was used as a starting point and then tests were done to identify that this particular value was good? I want to add a comment explaining its origin.

Reply:
Unfortunately I don't know precisely how `0x517cc1b727220a95` was found, I just tweaked what `rustc_hash` does (which contains that exact constant), which in turn documents it is a port of Firefox's internal hasher. You're right though that is related to pi, it seems the number for which `2^64 / N` is closest to pi.